### PR TITLE
Add fields to our log messages

### LINKED ISSUE #49: Add fields to our log messages
... instead of formatting the variables into the message text.

### PR DESCRIPTION
**Pull request by justinmueller05082000**
fixes #49

### CODE REVIEW

File: configobject/configsync/configsync.go [hunk 1]
```diff
@@ -57,13 +57,13 @@ func Operator(super *supervisor.Supervisor, chHA chan int, objectInformation *co
 		wgDelete     *sync.WaitGroup
 		wgUpdate     *sync.WaitGroup
 	)
-	log.Debugf("%s: Ready", objectInformation.ObjectType)
+	log.WithFields(log.Fields{"type" : objectInformation.ObjectType}).Info("Ready:")
```
Comment (Al2Klimov) [inline configobject/configsync/configsync.go]: Why `Info `?

File: configobject/configsync/configsync.go [hunk 2]
```diff
@@ -57,13 +57,13 @@ func Operator(super *supervisor.Supervisor, chHA chan int, objectInformation *co
...
 	for msg := range chHA {
 		switch msg {
 		// Icinga 2 probably restarted or died, stop operations and tell all workers to shut down.
 		case ha.Notify_StopSync:
 			if done != nil {
-				log.Debugf("%s: Lost responsibility", objectInformation.ObjectType)
+				log.WithFields(log.Fields{"type" : objectInformation.ObjectType}).Info("Lost responsibility:")
```
Comment (Al2Klimov) [inline configobject/configsync/configsync.go]: Why `Info `?

File: configobject/configsync/configsync.go [hunk 3]
```diff
@@ -73,7 +73,8 @@ func Operator(super *supervisor.Supervisor, chHA chan int, objectInformation *co
 				continue
 			}
 
-			log.Debugf("%s: Got responsibility", objectInformation.ObjectType)
+			log.WithFields(log.Fields{"type" : objectInformation.ObjectType}).Info("Got responsibility:")
+
```
Comment (Al2Klimov) [inline configobject/configsync/configsync.go]: Why `Info`?

Why the new line?

File: configobject/history/history.go [hunk 4]
```diff
@@ -448,8 +448,7 @@ func historyWorker(super *supervisor.Supervisor, historyType string, preparedSta
 	if len(entries) == 0 {
 		return
 	}
-
-	log.Debugf("%d %s history entries will be synced", len(entries), historyType)
+	log.WithFields(log.Fields{"amount" : len(entries), "type" : historyType}).Info("History entries will be synced")
```
Comment (Al2Klimov) [inline configobject/history/history.go]: Why `Info`?

File: configobject/history/history.go [hunk 5]
```diff
@@ -504,8 +503,9 @@ func historyWorker(super *supervisor.Supervisor, historyType string, preparedSta
 	historyCounter[historyType]++
 	historyCounterLock.Unlock()
 
-	log.Debugf("%d %s history entries synced", count, historyType)
-	log.Debugf("%d %s history entries broken", brokenEntries, historyType)
+	log.WithFields(log.Fields{"count" : count, "type": historyType}).Info("Following history entries synced")
+	log.WithFields(log.Fields{"broken_entries" : brokenEntries, "type": historyType}).Info("Following history entries broken")
+
```
Comment (Al2Klimov) [inline configobject/history/history.go]: Why `Info`?

Why the new line?

File: configobject/statesync/statesync.go [hunk 6]
```diff
@@ -80,8 +80,7 @@ func syncStates(super *supervisor.Supervisor, objectType string) {
 	if len(states) == 0 {
 		return
 	}
-
-	log.Debugf("%d %s state will be synced", len(states), objectType)
+	log.WithFields(log.Fields{"amount" : len(states), "type" : objectType}).Info("Going to be synced:")
```
Comment (Al2Klimov) [inline configobject/statesync/statesync.go]: Why `Info`?

File: configobject/statesync/statesync.go [hunk 7]
```diff
@@ -164,9 +163,8 @@ func syncStates(super *supervisor.Supervisor, objectType string) {
 
 	//Delete synced states from redis stream
 	super.Rdbw.XDel("icinga:state:stream:"+objectType, storedStateIds...)
-
-	log.Debugf("%d %s state synced", len(storedStateIds)-brokenStates, objectType)
-	log.Debugf("%d %s state broken", brokenStates, objectType)
+	log.WithFields(log.Fields{"amount" : len(storedStateIds)-brokenStates, "type" : objectType}).Info("State synced")
+	log.WithFields(log.Fields{"broken_states" : brokenStates, "type" : objectType}).Info("State broken")
```
Comment (Al2Klimov) [inline configobject/statesync/statesync.go]: Why `Info`?

Why the newline removal?

File: ha/ha.go [hunk 8]
```diff
@@ -136,7 +136,7 @@ func (h *HA) waitForEnvironment(chEnv chan *Environment) {
 func (h *HA) checkResponsibility() {
 	found, _, beat, err := h.getInstance()
 	if err != nil {
-		h.logger.Errorf("Failed to fetch instance: %v", err)
+		h.logger.WithFields(log.Fields{"error" : err}).Info("Failed to fetch instance")
```
Comment (Al2Klimov) [inline ha/ha.go]: Why `Info`?

File: ha/ha.go [hunk 9]
```diff
@@ -152,7 +152,7 @@ func (h *HA) checkResponsibility() {
 		}
 
 		if err != nil {
-			h.logger.Errorf("Failed to insert/update instance: %v", err)
+			h.logger.WithFields(log.Fields{"error" : err}).Info("Failed to insert/update instance")
```
Comment (Al2Klimov) [inline ha/ha.go]: Why `Info`?

File: ha/ha.go [hunk 10]
```diff
@@ -182,14 +182,14 @@ func (h *HA) runHA(chEnv chan *Environment) {
 			err := h.updateOwnInstance()
 
 			if err != nil {
-				h.logger.Errorf("Failed to update instance: %v", err)
+				h.logger.WithFields(log.Fields{"error" : err}).Info("Failed to update instance")
```
Comment (Al2Klimov) [inline ha/ha.go]: Why `Info`?

File: ha/ha.go [hunk 11]
```diff
@@ -182,14 +182,14 @@ func (h *HA) runHA(chEnv chan *Environment) {
...
 				h.super.ChErr <- errors.New("failed to update instance")
 				return
 			}
 		} else {
 			_, they, beat, err := h.getInstance()
 			if err != nil {
-				h.logger.Errorf("Failed to fetch instance: %v", err)
+				h.logger.WithFields(log.Fields{"error" : err}).Info("Failed to fetch instance")
```
Comment (Al2Klimov) [inline ha/ha.go]: Why `Info`?

File: ha/ha.go [hunk 12]
```diff
@@ -201,14 +201,14 @@ func (h *HA) runHA(chEnv chan *Environment) {
 				}
 
 				if err := h.updateOwnInstance(); err != nil {
-					h.logger.Errorf("Failed to update instance: %v", err)
+					h.logger.WithFields(log.Fields{"error": err}).Info("Failed to update instance")
```
Comment (Al2Klimov) [inline ha/ha.go]: Why `Info`?

File: ha/ha.go [hunk 13]
```diff
@@ -201,14 +201,14 @@ func (h *HA) runHA(chEnv chan *Environment) {
...
 					h.super.ChErr <- errors.New("failed to update instance")
 					return
 				}
 			} else if h.lastHeartbeat-beat > 15 {
 				h.logger.Info("Taking over.")
 				if err := h.takeOverInstance(); err != nil {
-					h.logger.Errorf("Failed to update instance: %v", err)
+					h.logger.WithFields(log.Fields{"error": err}).Info("Failed to update instance")
```
Comment (Al2Klimov) [inline ha/ha.go]: Why `Info`?

File: configobject/history/history.go [hunk 14]
```diff
@@ -169,7 +169,7 @@ func stateHistoryWorker(super *supervisor.Supervisor) {
 			stateType, err := strconv.ParseFloat(values["state_type"].(string), 32)
 
 			if err != nil {
-				log.Errorf("StateHistory: Could not parse stateType (%s) into float32", values["state_type"])
+				log.WithFields(log.Fields{"values" : values["state_type"]}).Info("StateHistory: Could not parse stateType into float32")
```
Comment (Al2Klimov) [inline configobject/history/history.go]: Why `Info`?

File: configobject/configsync/configsync.go [hunk 15]
```diff
@@ -57,13 +57,13 @@ func Operator(super *supervisor.Supervisor, chHA chan int, objectInformation *co
 		wgDelete     *sync.WaitGroup
 		wgUpdate     *sync.WaitGroup
 	)
-	log.Debugf("%s: Ready", objectInformation.ObjectType)
+	log.WithFields(log.Fields{"type" : objectInformation.ObjectType}).Debug("Ready:")
```
Comment (Al2Klimov) [inline configobject/configsync/configsync.go]: Why the colon inside the message?

File: configobject/configsync/configsync.go [hunk 16]
```diff
@@ -57,13 +57,13 @@ func Operator(super *supervisor.Supervisor, chHA chan int, objectInformation *co
...
 	for msg := range chHA {
 		switch msg {
 		// Icinga 2 probably restarted or died, stop operations and tell all workers to shut down.
 		case ha.Notify_StopSync:
 			if done != nil {
-				log.Debugf("%s: Lost responsibility", objectInformation.ObjectType)
+				log.WithFields(log.Fields{"type" : objectInformation.ObjectType}).Debug("Lost responsibility:")
```
Comment (Al2Klimov) [inline configobject/configsync/configsync.go]: Why the colon inside the message?

File: configobject/configsync/configsync.go [hunk 17]
```diff
@@ -73,7 +73,7 @@ func Operator(super *supervisor.Supervisor, chHA chan int, objectInformation *co
 				continue
 			}
 
-			log.Debugf("%s: Got responsibility", objectInformation.ObjectType)
+			log.WithFields(log.Fields{"type" : objectInformation.ObjectType}).Debug("Got responsibility:")
```
Comment (Al2Klimov) [inline configobject/configsync/configsync.go]: Why the colon inside the message?

File: configobject/history/history.go [hunk 18]
```diff
@@ -40,7 +40,7 @@ func logHistoryCounters() {
 		<-every20s.C
 		for _, historyType := range [6]string{"state", "notification", "usernotification", "downtime", "comment", "flapping"} {
 			if historyCounter[historyType] > 0 {
-				log.Infof("Added %d %s history entries in the last 20 seconds", historyCounter[historyType], historyType)
+				log.WithFields(log.Fields{"count" : historyCounter[historyType], "type" : historyType}).Info("History entries added in the last 20 seconds")
```
Comment (Al2Klimov) [inline configobject/history/history.go]: "Added some history entries" sound better than "History entries added".

File: configobject/history/history.go [hunk 19]
```diff
@@ -169,7 +169,7 @@ func stateHistoryWorker(super *supervisor.Supervisor) {
 			stateType, err := strconv.ParseFloat(values["state_type"].(string), 32)
 
 			if err != nil {
-				log.Errorf("StateHistory: Could not parse stateType (%s) into float32", values["state_type"])
+				log.WithFields(log.Fields{"values" : values["state_type"]}).Error("StateHistory: Could not parse stateType into float32")
```
Comment (Al2Klimov) [inline configobject/history/history.go]: "state_type" is a better field name than "values".

File: configobject/history/history.go [hunk 20]
```diff
@@ -504,8 +504,8 @@ func historyWorker(super *supervisor.Supervisor, historyType string, preparedSta
 	historyCounter[historyType]++
 	historyCounterLock.Unlock()
 
-	log.Debugf("%d %s history entries synced", count, historyType)
-	log.Debugf("%d %s history entries broken", brokenEntries, historyType)
+	log.WithFields(log.Fields{"count" : count, "type": historyType}).Debug("Following history entries synced")
+	log.WithFields(log.Fields{"broken_entries" : brokenEntries, "type": historyType}).Debug("Following history entries broken")
```
Comment (Al2Klimov) [inline configobject/history/history.go]: * "Following" => "Some"
* "broken_entries" => "count"

File: configobject/statesync/statesync.go [hunk 21]
```diff
@@ -53,7 +53,7 @@ func logSyncCounters() {
 	for {
 		<-every20s.C
 		if syncCounter["host"] > 0 || syncCounter["service"] > 0 {
-			log.Infof("Synced %d host and %d service states in the last 20 seconds", syncCounter["host"], syncCounter["service"])
+			log.WithFields(log.Fields{"sync_counter_host" : syncCounter["host"], "sync_counter_service" : syncCounter["service"]}).Info("Synced host and service states in the last 20 seconds")
```
Comment (Al2Klimov) [inline configobject/statesync/statesync.go]: * "Synced host" => "Synced some host"
* "sync_counter_host" => "host_states" and the same for the service

File: configobject/statesync/statesync.go [hunk 22]
```diff
@@ -80,8 +80,7 @@ func syncStates(super *supervisor.Supervisor, objectType string) {
 	if len(states) == 0 {
 		return
 	}
-
-	log.Debugf("%d %s state will be synced", len(states), objectType)
+	log.WithFields(log.Fields{"amount" : len(states), "type" : objectType}).Debug("Going to be synced:")
```
Comment (Al2Klimov) [inline configobject/statesync/statesync.go]: * "Some states will be synced" sounds better.
* Why the empty line removal?

File: configobject/statesync/statesync.go [hunk 23]
```diff
@@ -164,9 +163,8 @@ func syncStates(super *supervisor.Supervisor, objectType string) {
 
 	//Delete synced states from redis stream
 	super.Rdbw.XDel("icinga:state:stream:"+objectType, storedStateIds...)
-
-	log.Debugf("%d %s state synced", len(storedStateIds)-brokenStates, objectType)
-	log.Debugf("%d %s state broken", brokenStates, objectType)
+	log.WithFields(log.Fields{"amount" : len(storedStateIds)-brokenStates, "type" : objectType}).Debug("State synced")
+	log.WithFields(log.Fields{"broken_states" : brokenStates, "type" : objectType}).Debug("State broken")
```
Comment (Al2Klimov) [inline configobject/statesync/statesync.go]: * "broken_states" => "amount"
* Why the empty line removal?

File: prometheus/prometheus.go [hunk 24]
```diff
@@ -3,13 +3,14 @@
...
+	"fmt"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 	log "github.com/sirupsen/logrus"
 	"net/http"
 )
 
 func HandleHttp(addr string, chErr chan error) {
 	http.Handle("/metrics", promhttp.Handler())
-	log.Infof("Serving metrics at http://%s/metrics", addr)
+	log.WithFields(log.Fields{"address" : fmt.Sprintf("http://%s/metrics", addr)}).Info("Serving metrics at given website")
```
Comment (Al2Klimov) [inline prometheus/prometheus.go]: Remove "at given website".

File: prometheus/prometheus.go [hunk 25]
```diff
@@ -11,6 +11,7 @@ import (
 
 func HandleHttp(addr string, chErr chan error) {
 	http.Handle("/metrics", promhttp.Handler())
-	log.WithFields(log.Fields{"address" : fmt.Sprintf("http://%s/metrics", addr)}).Info("Serving metrics")
+	log.WithFields(log.Fields{
+		"address" : fmt.Sprintf("http://%s/metrics", addr)}).Info("Serving metrics")
```
Comment (Al2Klimov) [inline prometheus/prometheus.go]: What's missing here?

Comment (Al2Klimov) [reply prometheus/prometheus.go]: Or – even better – what's too much here?

Comment (justinmueller05082000) [reply prometheus/prometheus.go]: corrected it [here](https://github.com/Icinga/icingadb/pull/76/commits/49f2fd0ee621918d744d67f179f142873fe41f75)

Comment (Al2Klimov) [reply prometheus/prometheus.go]: OK, fine.

File: prometheus/prometheus.go [hunk 26]
```diff
@@ -3,13 +3,14 @@
...
+	"fmt"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 	log "github.com/sirupsen/logrus"
 	"net/http"
 )
 
 func HandleHttp(addr string, chErr chan error) {
 	http.Handle("/metrics", promhttp.Handler())
-	log.Infof("Serving metrics at http://%s/metrics", addr)
+	log.WithFields(log.Fields{"address" : fmt.Sprintf("http://%s/metrics", addr)}).Info("Serving metrics")
```
Comment (Al2Klimov) [inline prometheus/prometheus.go]: Please re-format the file.

**Comment by Al2Klimov** [comment]
## Blocked by

* [ ] #75

**Comment by N-o-X** [comment]
1) I'm not sure if those changes are really useful. I understand that the logs should be readable by tools like Graylog, but I don't think those log messages give enough context to those fields. 
2) I find it weird to always read "some" instead of the value and would prefer having both fields and values in the message. That way our logs are easily readable for humans and machines at the same time.

@justinmueller05082000 please wait before doing changes, I would like to hear what @bobapple says about this tomorrow.

**Comment by N-o-X** [comment]
After talking to @bobapple we decided on the following things:
- Human readable part: No fields, all information in msg
- Machine readable part: No msg, fields, formatted as JSON (logrus should be able to)
- We need a concrete logging concept for this

We need more time to talk about this and work on a concept. We'll to that after the RC.

**Comment by N-o-X** [comment]
This PR isn't needed anymore after our RC2 rework.

**Pull request closed without merging**